src/coreclr/jit/importercalls.cpp

```diff
@@ -3786,6 +3857,47 @@ GenTree* Compiler::impIntrinsic(CORINFO_CLASS_HANDLE    clsHnd,
                 break;
             }
 
+            case NI_System_ArgumentNullException_ThrowIfNull:
+            {
+                // void ThrowIfNull(object argument, string paramName = null)
+                // void ThrowIfNull(object argument, ExceptionArgument paramName)
+                assert(sig->numArgs == 2);
+                assert(sig->retType == CORINFO_TYPE_VOID);
+
+                if (opts.compDbgCode || opts.jitFlags->IsSet(JitFlags::JIT_FLAG_MIN_OPT))
+                {
+                    // Don't fold it for debug code or forced MinOpts
+                    break;
+                }
+
+                // if we see:
+                //
+                //   ArgumentNullException_ThrowIfNull(GT_BOX(...), ...)
+                //
+                // We should be able to remove the call (and the box). It is done via intrinsic
+                // because Tier0 is not able to remove the box itself (it doesn't inline callees).
+                //
+                GenTree* arg = impStackTop(1).val;
+                if (arg->OperIs(GT_BOX))
+                {
+                    impSpillSideEffects(true, CHECK_SPILL_ALL DEBUGARG("spill side effects for ThrowIfNull"));
+                    gtTryRemoveBoxUpstreamEffects(arg, BR_REMOVE_AND_NARROW);
+                    impPopStack();
+                    impPopStack();
+                    retNode = gtNewNothingNode();
+                }
```

Comment:
Would it be harder to handle both cases from `impThrowIfNull` to keep the handling in one place?

Reply:
Moved